clocks (#240)
monome/docs: norns/clocks.md

---
layout: default
parent: scripting
grand_parent: norns
title: clocks
nav_order: 3
permalink: /norns/clocks/
---

# clocks

the clock library provides a way to create timed function calls: loops, repetition, and delays. synchronization is possible and time base can come from a variety of sources. the clock library uses lua coroutines.

## commands

- `id = clock.run( func )` &mdash; start a new coroutine with function _func_. returns _id_
- `clock.cancel( id )` &mdash; cancel coroutine _id_ 
- `clock.sleep( time )` &mdash; resume in _time_ seconds
- `clock.sync( beats )` &mdash; resume in _beats_ according to tempo
- `beats = clock.get_time_beats()` &mdash; returns current time in _beats_
- `tempo = clock.get_tempo()` &mdash; returns current _tempo_
- `beat_sec = clock.get_beat_sec()` &mdash; returns length of beat in seconds

## callbacks

these functions can be script defined and are called by system events:

- `clock.transport.start` &mdash; called by transport start
- `clock.transport.stop` &mdash; called by transport stop

## use

### simple delay

> `clock.run` starts a new clock with the function you provide:

Reviewer: ```suggestion
`clock.run` starts a new coroutine with the function you provide:
```

```
function later()
  clock.sleep(2)
  print("now awake")
end

function init()
  print("starting now")
  clock.run(later)
  print("done with init")
end
```

when executed, you will see _starting now_, _done with init_, (two seconds delay and then): _now awake_. `clock.run` starts a coroutine with the function `later`. this function immediately sleeps for two seconds, which means the `init` function resumes.
   
### repetition and sync

you can sleep according to units of tempo (beats) by using `clock.sync` instead of `clock.sleep`.

```
engine.name = 'PolyPerc'

function strum()
  for i=1,8 do
    engine.hz(i*100)
    clock.sync(1/4)
  end
end

function key(n,z)
  if n==3 and z==1 then
    clock.run(strum)
  end
end
```

> this script creates a new clock when K3 is pressed. the `strum` function plays eight notes, sleeping for 1/4 of a beat between notes.

Reviewer: ```suggestion
this script starts a new coroutine when K3 is pressed. the `strum` function plays eight notes, sleeping for 1/4 of a beat between notes.
```

```
engine.name = 'PolyPerc'

function strum(n,speed)
  for i=1,n do
    clock.sync(1/speed)
    engine.hz(i*100)
  end
end

function key(n,z)
  if n==3 and z==1 then
    clock.run(strum, math.random(16), math.random(8))
  end
end
```

now each `strum` is executed with different arguments. note that you can overlap multiple runs by pressing the key rapidly! a new coroutine is created with each run, so you can have numerous clock processes running at once.

`clock.sync` sleeps until the next subdivision specified arrives, so the timing is effectively quantized to the master tempo (which is the goal of this use case).

### loop

so far all of our clock functions end by themselves, by reaching the end of their own execution. we can create forever-running loops using a `while true` inner loop. these will continue running until you cancel them manually with `clock.cancel`, or by launch a new script.

```
engine.name = 'PolyPerc'

clock_id = 0

function forever(freq)
  while true do
    clock.sync(1/4)
    engine.hz(freq)
  end
end

function init()
  clock_id = clock.run(forever, 220)
end

function key(n,z)
  if n==3 and z==1 then
    clock.cancel(clock_id)
  end
end
```

> to cancel a clock you need to save its `id`.

Reviewer: ```suggestion
to cancel a coroutine you need to save its `id`.
```

### multiples

> you can run lots of clocks at once: up to 100.

Reviewer: ```suggestion
you can run lots of coroutines at once: up to 100.
```

---
layout: default
parent: scripting
grand_parent: norns
title: clocks
nav_order: 3
permalink: /norns/clocks/
---

# clocks

the clock library provides a way to create timed function calls: loops, repetition, and delays. synchronization is possible and time base can come from a variety of sources.

## commands

> - `id = clock.run( func )` &mdash; start a new clock with function _func_. returns _id_

Reviewer: ```suggestion
- `id = clock.run( func )` &mdash; start a new coroutine with function _func_. returns _id_
```

> - `id = clock.run( func )` &mdash; start a new clock with function _func_. returns _id_

Reviewer: start a new clock or coroutine?

> - `clock.cancel( id )` &mdash; cancel clock _id_ 

Reviewer: same, is coroutine implied here?

- `clock.sleep( time )` &mdash; resume in _time_ seconds

> - `clock.sync( beats )` &mdash; resume in _beats_ according to tempo

Reviewer: sync doesn't exactly resume _in_ `beats` beats, but rather when `beats` beats are reached on the timeline, e.g. if you call `clock.sync(3)` around beat 2 according to the clock, coroutine will wake up in 1 beat.

upd: i see that it's covered below, so is probably fine to keep here as is.

Reviewer 2: i think this could be clarified -- it does sound like `clock.sync(3)` would resume in 3 beats from whatever beat its called, rather than resume on beat 3 according to the global clock.

> - `clock.sync( beats )` &mdash; resume in _beats_ according to tempo

Reviewer 2: ```suggestion
- `clock.sync( beats )` &mdash; resume at _beats_ count according to global tempo
```

- `beats = clock.get_time_beats()` &mdash; returns current time in _beats_
- `tempo = clock.get_tempo()` &mdash; returns current _tempo_

> - `beat_sec = clock.get_beat_sec()` &mdash; returns length of beat in seconds

Reviewer 2: ```suggestion
- `beat_sec = clock.get_beat_sec()` &mdash; returns length of a single beat at current tempo in seconds
```

## callbacks

these functions can be script defined and are called by system events:

- `clock.transport.start` &mdash; called by transport start
- `clock.transport.stop` &mdash; called by transport stop

## use

### simple delay

`clock.run` starts a new clock with the function you provide:

```

> function later()
>   clock.sleep(2)
>   print("now awake")
> end
> 
> function init()
>   print("starting now")
>   clock.run(later)
>   print("done with init")
> end

Reviewer 2: ```suggestion
function init()
  print("starting now")
  clock.run(later)
  print("done with init")
end

function later()
  clock.sleep(2)
  print("now awake")
end
```

reads a bit more linearly to see `clock.run` (the first thing a reader is told to look for in this section) before `later()`.

```

> when executed, you will see _starting now_, _done with init_, (two seconds delay and then): _now awake_. `clock.run` starts a coroutine with the function `later`. this functions immediately sleeps for two seconds, which means the `init` function resumes.

Reviewer 2: ```suggestion
when executed, you will see _starting now_, _done with init_, (two seconds delay and then): _now awake_. `clock.run` starts a coroutine with the function `later`. this function immediately sleeps for two seconds, which means the `init` function resumes and finishes before `later` is able to print its output.
```

>    

Reviewer 3: this functions immediately sleeps =>

this function immediately sleeps

### repetition and sync

you can sleep according to units of tempo (beats) by using `clock.sync` instead of `clock.sleep`.

```
engine.name = 'PolyPerc'

function strum()
  for i=1,8 do

>     engine.hz(i*100)
>     clock.sync(1/4)

Reviewer 2: ```suggestion
    clock.sync(1/4)
    engine.hz(i*100)
```

unless we want to show a situation where the first note is un-quantized, this will make sure all notes are synced to 1/16th tick.

  end
end

function key(n,z)
  if n==3 and z==1 then
    clock.run(strum)
  end
end
```

> this script creates a new clock when K3 is pressed. the `strum` function plays eight notes, sleeping for 1/4 of a beat between notes.

Reviewer 2: ```suggestion
this script creates a new clock when K3 is pressed. the `strum` function plays eight notes, sleeping for 1/4 of a beat (= 1/16ths in 4/4) between notes, synced to the global tempo.
```

you can change the global tempo using parameters via command or menu. see the later section.

### arguments

pass arguments to a clock function by appending them within the `clock.run` call:

```
engine.name = 'PolyPerc'

function strum(n,speed)
  for i=1,n do
    clock.sync(1/speed)
    engine.hz(i*100)
  end
end

function key(n,z)
  if n==3 and z==1 then
    clock.run(strum, math.random(16), math.random(8))
  end
end
```

> now each `strum` is executed with different agruments. note that you can overlap multiple runs by pressing the key rapidly! a new coroutine is created with each run, so you can have numerous clock processes running at once.

Reviewer 3: agruments =>

arguments

> `clock.sync` sleeps until the next subdivision specified arrives, so the timing is effectively quantized to the master tempo (which is the goal of this use case).

Reviewer 2: ```suggestion
now each `strum` is executed with different arguments. note that you can overlap multiple runs by pressing the key rapidly! a new coroutine is started with each run, so you can have numerous processes running at once.

`clock.sync` sleeps until the next specified subdivision arrives, so the timing is effectively quantized to the global tempo (which is the goal of this use case).
```

### loop

> so far all of our clock functions end by themselves, by reaching the end of their own execution. we can create forever-running loops using a `while(true)` inner loop. these will continue running until you cancel them manually with `clock.cancel`, or by launch a new script.

Reviewer: ```suggestion
so far all of our clock functions end by themselves, by reaching the end of their own execution. we can create forever-running loops using a `while true` inner loop. these will continue running until you cancel them manually with `clock.cancel`, or by launch a new script.
```
(in lua parentheses aren't necessary for while loop)

```
engine.name = 'PolyPerc'

clock_id = 0

function forever(freq)

>   while(true) do

Reviewer: ```suggestion
  while true do
```

    clock.sync(1/4)
    engine.hz(freq)
  end
end

function init()
  clock_id = clock.run(forever, 220)
end

function key(n,z)
  if n==3 and z==1 then
    clock.cancel(clock_id)
  end
end
```

to cancel a clock you need to save its `id`.

### multiples

you can run lots of clocks at once: up to 100.

```
engine.name = 'PolyPerc'

function forever(freq,rate)

>   while(true) do

Reviewer: ```suggestion
  while true do
```

## commands

- `id = clock.run( func )` &mdash; start a new coroutine with function _func_. returns _id_
- `clock.cancel( id )` &mdash; cancel coroutine _id_ 
- `clock.sleep( time )` &mdash; resume in _time_ seconds
- `clock.sync( beats )` &mdash; resume at _beats_ count according to global tempo
- `beats = clock.get_time_beats()` &mdash; returns current time in _beats_
- `tempo = clock.get_tempo()` &mdash; returns current _tempo_
- `beat_sec = clock.get_beat_sec()` &mdash; returns length of a single beat at current tempo in seconds

## callbacks

these functions can be script defined and are called by system events:

- `clock.transport.start` &mdash; called by transport start
- `clock.transport.stop` &mdash; called by transport stop

## use

### simple delay

`clock.run` starts a new coroutine with the function you provide:

```
function later()
  clock.sleep(2)
  print("now awake")
end

function init()
  print("starting now")
  clock.run(later)
  print("done with init")
end
```

> when executed, you will see _starting now_, _done with init_, (two seconds delay and then): _now awake_. `clock.run` starts a coroutine with the function `later`. this function immediately sleeps for two seconds, which means the `init` function resumes.

Reviewer 2: ```suggestion
when executed, you will see _starting now_, _done with init_, (two seconds delay and then): _now awake_. `clock.run` starts a coroutine with the function `later`. this function immediately sleeps for two seconds, which means the `init` function resumes and finishes before `later` is able to print its output.
```

   
### repetition and sync

you can sleep according to units of tempo (beats) by using `clock.sync` instead of `clock.sleep`.

```
engine.name = 'PolyPerc'

function strum()
  for i=1,8 do
    engine.hz(i*100)
    clock.sync(1/4)
  end
end

function key(n,z)
  if n==3 and z==1 then
    clock.run(strum)
  end
end
```

> this script starts a new coroutine when K3 is pressed. the `strum` function plays eight notes, sleeping for 1/4 of a beat between notes.

Reviewer 2: ```suggestion
this script starts a new coroutine when K3 is pressed. the `strum` function plays eight notes, sleeping for 1/4 of a beat (= 1/16th note in 4/4) between notes, synced to the global tempo.
```

you can change the global tempo using parameters via command or menu. see the later section.

### arguments

pass arguments to a clock function by appending them within the `clock.run` call:

```
engine.name = 'PolyPerc'

function strum(n,speed)
  for i=1,n do
    clock.sync(1/speed)
    engine.hz(i*100)
  end
end

function key(n,z)
  if n==3 and z==1 then
    clock.run(strum, math.random(16), math.random(8))
  end
end
```

> now each `strum` is executed with different arguments. note that you can overlap multiple runs by pressing the key rapidly! a new coroutine is created with each run, so you can have numerous clock processes running at once.
> 
> `clock.sync` sleeps until the next subdivision specified arrives, so the timing is effectively quantized to the master tempo (which is the goal of this use case).

Reviewer 2: ```suggestion
now each `strum` is executed with different arguments. note that you can overlap multiple runs by pressing the key rapidly! a new coroutine is started with each run, so you can have numerous processes running at once.

`clock.sync` sleeps until the next subdivision specified arrives, so the timing is effectively quantized to the global tempo (which is the goal of this use case).
```

### loop

> so far all of our clock functions end by themselves, by reaching the end of their own execution. we can create forever-running loops using a `while true` inner loop. these will continue running until you cancel them manually with `clock.cancel`, or by launch a new script.

Reviewer 2: ```suggestion
so far all of our coroutines end by themselves, by reaching the end of their own execution. we can create forever-running loops using a `while true` inner loop. these will continue running until you launch a new script or cancel them manually with `clock.cancel`.

to cancel a coroutine, you need to store its `id`. in the following script, we establish a global variable `clock_id` to capture the identifier for our running function, so that we can directly cancel it later.
```

```
engine.name = 'PolyPerc'

> clock_id = 0

Reviewer 2: this gets overwritten in the `init()` -- i think it actually confuses things to have the clock_id explicitly defined here since executing `clock.cancel(0)` won't cancel the script's clock.

Author: the problem is we need the global variable `clock_id` created outside of `init` which means assigning it to a default value. would it make more sense to have it be `clock_id = "not assigned"`? because yes, we are trying to show it gets overwritten

Reviewer: value of nil will work fine, clock.cancel(nil) will be silently ignored

Reviewer: same (nothing) will happen for clock.cancel(0)

Author: yes, but then the code won't do what it's supposed to do, which is capture the id of the clock assigned in init, and then cancel it with a key?

Reviewer 2: "because yes, we are trying to show it gets overwritten"
ahhh gotcha. let's just make that a bit more explicit in the words?

Author: comment added :)

function forever(freq)
  while true do
    clock.sync(1/4)
    engine.hz(freq)
  end
end

function init()
  clock_id = clock.run(forever, 220)
end

function key(n,z)
  if n==3 and z==1 then
    clock.cancel(clock_id)
  end
end
```

> to cancel a coroutine you need to save its `id`.

Reviewer 2: ```suggestion

```
covered ahead of the script.

### multiples

> you can run lots of coroutines at once: up to 100.

Reviewer 2: ```suggestion
you can run lots of coroutines at once: up to 100.

in this script, we capture `id`s using a table of `voice`s and print the assignments to the REPL. you'll notice that the `id`s change (by incrementing) each time -- but by using variables to capture the `id`, we have a predictable way to terminate them.
```

```

> engine.name = 'PolyPerc'
> 
> function forever(freq,rate)
>   while true do
>     clock.sync(1/rate)
>     engine.hz(freq)
>   end
> end
> 
> function init()
>   clock.run(forever,333,3)
>   clock.run(forever,666,1)
>   clock.run(forever,999,2)
>   clock.run(forever,111,0.33)
> end

Reviewer 2: ```suggestion
engine.name = 'PolyPerc'

function forever(freq,rate)
  while true do
    clock.sync(1/rate)
    engine.hz(freq)
  end
end

function init()
  voice = {1,2,3,4}
  voice[1] = clock.run(forever,333,3)
  voice[2] = clock.run(forever,666,1)
  voice[3] = clock.run(forever,999,2)
  voice[4] = clock.run(forever,111,0.33)
  
  for i = 1,4 do
    print("voice "..i..": ID "..voice[i])
  end
end
```

> you can run lots of coroutines at once: up to 100.
> 
> ```
> engine.name = 'PolyPerc'
> 
> function forever(freq,rate)
>   while true do
>     clock.sync(1/rate)
>     engine.hz(freq)
>   end
> end
> 
> function init()
>   clock.run(forever,333,3)
>   clock.run(forever,666,1)
>   clock.run(forever,999,2)
>   clock.run(forever,111,0.33)
> end
> ```

Reviewer 2: ```suggestion
you can run lots of coroutines at once: up to 100.

in this script, we capture `id`s using a table of `voice`s and print the assignments to the REPL. you'll notice that the `id`s change (by incrementing) each time -- but by using variables to capture the `id`, we have a predictable way to terminate them.

```
engine.name = 'PolyPerc'

function forever(freq,rate)
  while true do
    clock.sync(1/rate)
    engine.hz(freq)
  end
end

function init()
  voice = {1,2,3,4}
  voice[1] = clock.run(forever,333,3)
  voice[2] = clock.run(forever,666,1)
  voice[3] = clock.run(forever,999,2)
  voice[4] = clock.run(forever,111,0.33)
  
  for i = 1,4 do
    print("voice "..i..": ID "..voice[i])
  end
end
```

use the maiden command line to cancel voices, eg. `clock.cancel(voice[3])`

> 

Reviewer 2: ```suggestion
use maiden's command line to cancel individual voices, eg `clock.cancel(voice[3])`

```
